File: ts/data/omitted-sample.ts
import { TableDefinition } from "../component/table-builder";
import { Tooltip } from "../component/tooltip";
import { makeNameDiv, makeTextDivWithTooltip } from "../util/html-utils";
import { siteConfig } from "../util/site-config";
import { urls } from "../util/urls";
import { Donor } from "./case";

interface OmittedSample {
  id: string;
  name: string;
  requisitionId?: number;
  requisitionName?: string;
  assayId?: number;
  project: string;
  donor: Donor;
  createdDate: string;
}

export const omittedSampleDefinition: TableDefinition<OmittedSample, void> = {
  queryUrl: urls.rest.omissions,
  defaultSort: {
    columnTitle: "Created Date",
    descending: true,
    type: "date",
  },
  filters: [
    {
      title: "Donor",
      key: "DONOR",
      type: "text",
      autocompleteUrl: urls.rest.autocomplete.donorNames,
    },
    {
      title: "Project",
      key: "PROJECT",
      type: "text",
      autocompleteUrl: urls.rest.autocomplete.projectNames,
    },
    {
      title: "Requisition",
      key: "REQUISITION",
      type: "text",
      autocompleteUrl: urls.rest.autocomplete.requisitionNames,
    },
  ],
  generateColumns(data) {
    return [
      {
        title: "Name",
        sortType: "text",
        addParentContents(sample, fragment) {
          fragment.appendChild(
            makeNameDiv(sample.name, urls.miso.sample(sample.id))
          );
        },
      },
      {
        title: "Project",
        sortType: "text",
        addParentContents(sample, fragment) {
          fragment.appendChild(
            makeNameDiv(
              sample.project,
              urls.miso.project(sample.project),
              urls.dimsum.project(sample.project)
            )
          );
        },
      },
      {
        title: "Donor",
        sortType: "text",
        addParentContents(sample, fragment) {
          fragment.appendChild(
            makeNameDiv(
              sample.donor.name,
              urls.miso.sample(sample.donor.id),
              urls.dimsum.donor(sample.donor.name)
            )
          );
          fragment.appendChild(
            makeTextDivWithTooltip(sample.donor.externalName, "External Name")
          );
        },
      },
      {
        title: "Requisition",
        sortType: "text",
        addParentContents(sample, fragment) {
          if (sample.requisitionId && sample.requisitionName) {
            fragment.appendChild(
              makeNameDiv(
                sample.requisitionName,
                urls.miso.requisition(sample.requisitionId)
              )
            );
            if (sample.assayId) {
              const assay = siteConfig.assaysById[sample.assayId];
              const assayDiv = document.createElement("div");
              assayDiv.innerText = assay.name;
              fragment.appendChild(assayDiv);
            }
          }
        },
      },
      {
        title: "Created Date",
        sortType: "date",
        addParentContents(sample, fragment) {
          fragment.appendChild(document.createTextNode(sample.createdDate));
        },
      },
    ];
  },
};

File: src/main/java/ca/on/oicr/gsi/dimsum/data/OmittedSample.java
package ca.on.oicr.gsi.dimsum.data;

import static java.util.Objects.requireNonNull;

import java.time.LocalDate;

public class OmittedSample {

  private final String id;
  private final String name;
  private final Long requisitionId;
  private final String requisitionName;
  private final Long assayId;
  private final String project;
  private final Donor donor;
  private final LocalDate createdDate;

  private OmittedSample(Builder builder) {
    this.id = requireNonNull(builder.id);
    this.name = requireNonNull(builder.name);
    this.requisitionId = builder.requisition == null ? null : builder.requisition.getId();
    this.requisitionName = builder.requisition == null ? null : builder.requisition.getName();
    this.assayId = builder.requisition == null ? null : builder.requisition.getAssayId();
    this.project = requireNonNull(builder.project);
    this.donor = requireNonNull(builder.donor);
    this.createdDate = requireNonNull(builder.createdDate);
  }

  public String getId() {
    return id;
  }

  public String getName() {
    return name;
  }

  public Long getRequisitionId() {
    return requisitionId;
  }

  public String getRequisitionName() {
    return requisitionName;
  }

  public Long getAssayId() {
    return assayId;
  }

  public String getProject() {
    return project;
  }

  public Donor getDonor() {
    return donor;
  }

  public LocalDate getCreatedDate() {
    return createdDate;
  }

  public static class Builder {

    private String id;
    private String name;
    private Requisition requisition;
    private String project;
    private Donor donor;
    private LocalDate createdDate;

    public Builder id(String id) {
      this.id = id;
      return this;
    }

    public Builder name(String name) {
      this.name = name;
      return this;
    }

    public Builder requisition(Requisition requisition) {
      this.requisition = requisition;
      return this;
    }

    public Builder project(String project) {
      this.project = project;
      return this;
    }

    public Builder donor(Donor donor) {
      this.donor = donor;
      return this;
    }

    public Builder createdDate(LocalDate createdDate) {
      this.createdDate = createdDate;
      return this;
    }

    public OmittedSample build() {
      return new OmittedSample(this);
    }

  }


}
